hd-wallet/pluginv2.lua

```diff
@@ -0,0 +1,452 @@
+        child_key.version = PUBLIC_WALLET_VERSION
+        fingerprint = hash160(parent_key.key)
+        child_key.fingerprint = string.sub(fingerprint, 1, 8)
+        key_bytes = public_key_for_private_key(string.sub(hmac, 1, 64))
+        child_key.key = add_public_keys(key_bytes, parent_key.key)
+    end
+
+   -- checksum: double sha256 of serialized key
+    local child_key_string = child_key.version.. child_key.depth..  child_key.fingerprint.. child_key.index.. child_key.chain_code.. child_key.key
+    local sha256_hash1 = assert(digest { data = Blob.from_hex(child_key_string), alg = 'SHA256' }).digest:hex()
+    local sha256_hash2 = assert(digest { data =  Blob.from_hex(sha256_hash1), alg = 'SHA256' }).digest:hex()
+    child_key.checksum = string.sub(sha256_hash2, 1, 8)
+
+    return child_key
+end
+
+----------- @@@@@@@@@@@@@@@@@@@@@@@@@ -----------
+----------- Main method for ETH       -----------
+----------- @@@@@@@@@@@@@@@@@@@@@@@@@ -----------
+function sign_eth(input)
+    local exported_master_key_serialized = export_secret_key(input.masterKeyId)
+    local master_key = deserialize(exported_master_key_serialized:bytes())
+
+    local private_key = string.sub(master_key.key, 3, 66)
+
+    -- import private key as asn1 ec key
+    local asn1_ec_key = create_ASN1_private_key(private_key)
+    local blob = Blob.from_hex(asn1_ec_key)
+    local ec_child_key = import_ec_key(blob)
+
+    local signature = assert(ec_child_key:sign { hash = Blob.from_hex(input.msgHash), hash_alg = "SHA256", deterministic_signature = true }).signature
+
+    return {
+      coin = input.coin,
+      signature = signature:hex():lower()
+    }
+end
+
+----------- @@@@@@@@@@@@@@@@@@@@@@@@@ -----------
+----------- Main method for XRP       -----------
+----------- @@@@@@@@@@@@@@@@@@@@@@@@@ -----------
+function sign_xrp(input)
+    local exported_master_key_serialized = export_secret_key(input.masterKeyId)
+    local master_key = deserialize(exported_master_key_serialized:bytes())
+
+    local private_key = string.sub(master_key.key, 3, 66)
+
+    -- import private key as asn1 ec key
+    local asn1_ec_key = create_ASN1_private_key(private_key)
+    local blob = Blob.from_hex(asn1_ec_key)
+    local ec_child_key = import_ec_key(blob)
+
+    local signature = assert(ec_child_key:sign { hash = Blob.from_hex(input.msgHash), hash_alg = "SHA256", deterministic_signature = true }).signature
+
+    return {
+      coin = input.coin,
+      signature = signature:hex():lower()
+    }
+end
+
+----------- @@@@@@@@@@@@@@@@@@@@@@@@@ -----------
+----------- Main method for UTXO      -----------
+----------- @@@@@@@@@@@@@@@@@@@@@@@@@ -----------
+function sign_utxo(input)
+    local exported_master_key_serialized = export_secret_key(input.masterKeyId)
+    local master_key = deserialize(exported_master_key_serialized:bytes())
+    local indices = parse_path(input.path)
+
+    for i = 2, #indices do
+        child_key = derive_new_child(master_key, tonumber(indices[i]))
```

Comment:
This actually creates a global variable `child_key`. 
It means that everywhere else, `child_key` needs to be defined `local`ly, which is a bit error prone. I suggest to define it here locally (`local child_key` before the `for` loop, for instance).

(Apart from this one in particular, there are other global variables in the plugin)